jobsprofiler: dump trace recording on job completion

adityamaru opened the pull request:
This change teaches the job resumer to fetch and write its trace recording before finishing its tracing span. These traces will be consumed by the job profiler bundle that is being introduced in #105384. These traces will be valuable in understanding a job's execution characteristics during each resumption, even if the job has reached a terminal state.

Currently, this behaviour is opt-in and has been enabled for backups, restore, import and physical replication jobs.

Informs: #102794
Release note: None

adityamaru committed:
jobsprofiler: dump trace recording on job completion

This change teaches the job resumer to fetch and write
its trace recording before finishing its tracing span.
These traces will be a part of the execution detail files
introduced in #105384. These traces will be
valuable in understanding a job's execution characteristics
during each resumption, even if the job has reached a terminal
state.

Currently, this behaviour is opt-in and has been enabled for
backups, restore, import and physical replication jobs.

Informs: #102794
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105624)
<!-- Reviewable:end -->

adityamaru commented:
friendly ping @dt @stevendanna

adityamaru committed:
sql: refactor execution detail utility methods

This change is a refactor only commit to move some of
the utility methods used to read, write and list execution
detail files to `pkg/jobs`. This allows us to use these
methods without pulling in a dependancy to the sql package.

Release note: None

adityamaru commented:
TFTR!

bors r=dt
